### ISSUE TITLE
On processing long audio files / signals

### ISSUE THREAD

**Issue by ejhumphrey** [librosa/librosa#772]
`<dislaimer>` There are two levels to this possible enhancement discussion, and since this could hypothetically result in implementation changes, figured it made sense here rather than google group – if that's wrong, happy to redirect! `</disclaimer>`

To date, we've gotten pretty far with loading full signals into memory and transforming them without issue, because it's 2018 and everyone has GBs of headroom to play with. Music files, generally, aren't that long, and that makes things easy. 

I'm curious, though, what the preferred course of action would be for, say, hours of audio, when you start to run into resource limitations. There may be others, but I see two kinds of memory issues that crop up:
- the input signal / output signal doesn't fit in memory
- some _intermediary_ representation doesn't fit in memory, but the input & output do, e.g. a matrix product before reduction

There are some simple, albeit unsatisfying, options:
- slice a long recording up into smaller bits at the filesystem level
- process the file in chunks in Python scope

While both are simple in principle, they introduce a layer of bookkeeping I (as a user) would love to not have to think about. I see a few ways to engineer around this, and there are no doubt other (better?) ones:
- A file in / file out API
- buffered iterators and reducers

I suppose a preliminary question is, "is this even `librosa`'s problem?" I'm asking here, so I think it at least could be, but would be keen to kick around thoughts? (and also help implement a solution, if a good one arises).

**Comment by bmcfee** [comment]
The TLDR here is that both options (file streaming and generators) are viable, and we're working on providing an abstraction layer to make it easy to use.  Of course, not everything that librosa does will be compatible with this mode: resampling (and CQTs) use non-causal interpolation filters; recurrence features are non-local, beat tracking is a global DP, etc etc.  But for basic feature analysis, we should have a good streaming solution within the next few months.  If you have any interesting use patterns, it'd be great to see that earlier rather than later to help guide our design.

Meanwhile, you might want to check out [this page](https://librosa.github.io/librosa/ioformats.html#blockwise-reading) of the docs on doing streaming feature extraction using soundfile.

**Comment by ejhumphrey** [comment]
ah, thanks for the redirect ... I'd searched issues for something similar, but didn't quickly see anything that matched. Guess this (at present) is kind of a usage question, though it's good to know that an abstraction layer is in the works. I don't really have any more guidance around use-case besides "so I have a 120 minute file".

Tangent: It's interesting to consider that some kind of boundary finding might make for good cut-points, so that non-causal processing could be applied to those regions independently. This is wandering off into a distinct algorithm altogether, but figured it's worth mentioning in passing.

**Comment by bmcfee** [comment]
Following up on this, #864 provides an example (in the gallery) of how to use soundfile block streaming with STFT and pcen (which is causal, and now allows you to carry over the final state / initialize on the next call).

I mention this here because there's a subtle detail in getting the block streaming to exactly match buffered reading, and it's perhaps worth abstracting that out into its own helper function.

Soundfile lets you read a specified length at a time, and provides an option to rewind by some `n` samples when generating the next block; that is, it lets you do overlapping blocks.  While this in theory would be great for doing framewise analysis directly, it turns out to be quite inefficient in practice.  What works better is to use relatively long blocks (eg 30 seconds or more, assuming latency is not a factor here), disable frame centering/padding, and use the rewind of `n` to overlap the last frame of one block with the first frame of the next block.

This gets pretty delicate, and depends a lot on what specific analysis will be done downstream, but it might be worth trying out a few more prototypes and seeing if there's some nice helper functionality that we can pull out.  Something like:

```python
for block in librosa.stream(filename, frame_length=2048, hop_length=512, block_length=2048, ...):
    [your code here]
```
that would handle the block / frame / hop parameters in a sane way.  I'm envisioning this as a pure generator, but it might be necessary to have a more OOPy approach here to expose sampling rate, etc parameters.
